qutip/nonmarkov/bofin.py

```diff
@@ -0,0 +1,1553 @@
+        data = 0.5 * (data + data.H)
+
+        solution = solution.reshape((nstates, self.H_sys.shape[0] ** 2))
+
+        return Qobj(data, dims=dims), solution
+
+    def run(self, rho0, tlist, full_init=False, return_full=False):
+        """
+        Function to solve the time dependent evolution of the ODE given
+        an initial condition and set of time steps.
+
+        Parameters
+        ----------
+        rho0 : Qobj
+            Initial state (density matrix) of the system
+            (if full_init==False).
+            If full_init = True, then rho0 should be a numpy array of
+            initial state and all ADOs.
+
+        tlist : list
+            Time over which system evolves.
+
+        full_init: Boolean
+            Indicates if initial condition is just the system Qobj, or a
+            numpy array including all ADOs.
+
+        return_full: Boolean
+
+            Whether to also return as output the full state of all ADOs.
+
+        Returns
+        -------
+        results : :class:`qutip.solver.Result`
+            Object storing all results from the simulation.
+            If return_full == True, also returns ADOs as an additional
+            numpy array.
+        """
+
+        sup_dim = self._sup_dim
+
+        solver = self._ode
+        dims = self.coup_op[0].dims
+        shape = self.coup_op[0].shape
+
+        if not self._configured:
+            raise RuntimeError("Solver must be configured before it is run")
+
+        output = Result()
+        output.solver = "hsolve"
+        output.times = tlist
+        output.states = []
+        if not full_init:
+            output.states.append(Qobj(rho0))
+            rho0_flat = rho0.full().ravel('F')
+            rho0_he = np.zeros([sup_dim*self.nhe], dtype=complex)
+            rho0_he[:sup_dim] = rho0_flat
+            solver.set_initial_value(rho0_he, tlist[0])
+        else:
+            output.states.append(Qobj(
+                rho0[:sup_dim].reshape(shape, order='F'), dims=dims,
+            ))
+            rho0_he = rho0
+            solver.set_initial_value(rho0_he, tlist[0])
+
+        dt = np.diff(tlist)
+        n_tsteps = len(tlist)
+
+        if not return_full:
+            self.progress_bar.start(n_tsteps)
+            for t_idx, t in enumerate(tlist):
+                self.progress_bar.update(t_idx)
+                if t_idx < n_tsteps - 1:
+                    solver.integrate(solver.t + dt[t_idx])
+                    rho = Qobj(
+                        solver.y[:sup_dim].reshape(shape, order="F"),
+                        dims=dims
+                    )
+                    output.states.append(rho)
+
+            self.progress_bar.finished()
+            return output
+
+        else:
+            self.progress_bar.start(n_tsteps)
+            N_he = self.nhe
+            N = shape[0]
+            hshape = (N_he, N**2)
+            full_hierarchy = [rho0.reshape(hshape)]
```

Comment:
minor comment:  this reshape is useful for the utility functions in the example notebooks, for extracting bath properties like current and things. But it is a little inconsistent with how full__init=True is used, where the assumed shape is different. Maybe set this reshape as an option? (if so, the same should be done to the fermion class)